src/register/pmpaddrx.rs

```diff
@@ -22,47 +27,48 @@ impl PmpAddr {
     }
 
     #[inline]
-    pub fn encode(&mut self, mode: Mode, addr: usize, size: Option<NonZeroUsize>) {
+    pub fn encode(&mut self, mode: Mode, addr: Addr, size: Option<NonZeroSize>) {
         self.bits = match mode {
             Mode::OFF => 0,
-            Mode::TOR => addr >> 2,
-            Mode::NA4 => addr >> 2,
+            Mode::TOR => (addr >> 2).try_into().unwrap(),
+            Mode::NA4 => (addr >> 2).try_into().unwrap(),
             Mode::NAPOT => Self::encode_napot(addr, size.unwrap().into()),
         }
     }
 
     #[inline]
-    fn encode_napot(addr: usize, size: usize) -> usize {
+    fn encode_napot(addr: Addr, size: Size) -> usize {
         // See riscv priv spec "Physical Memory Protection CSRs
         // "Each PMP address register encodes bits 33–2 of a 34-bit physical address for RV32"
         // and
         // "For RV64, each PMP address register encodes bits 55–2 of a 56-bit physical address"
-        // TODO: top bits will get lost on 64bit system
-        let addr = addr >> 2;
+        let addr: usize = (addr >> 2).try_into().unwrap();
 
         let mut pmpaddr: usize = 0;
         pmpaddr |= addr;
-        pmpaddr |= (size - 1) >> 3;
+        // verify the provided size is valid
+        assert!(((size - 1) >> 3) <= usize::MAX as Size);
```

Comment:
Shouldn't there be a check to ensure that `(address & ((size - 1) >> 3))` equals 0 i.e. address doesn't have any lower bits set where size will be ORed? Also, do we need a check to ensure that the provided size is a power of 2?

Reply:
Hmm, this library didn't really do input validation before, but it seemed wrong to downcast like that without checking.   I will go through and add more complete checks.  I also think I will change the return types to results since there will be more ways to fail.

Reply:
I added some checks.  Take a look and let me know if I missed anything.